File: 207/course_schedule.go
// 207. Course Schedule
// https://leetcode.com/problems/course-schedule/
package main

import (
	"github.com/k0kubun/pp"
)

// May LeetCoding Challenge Day 29
// https://leetcode.com/explore/challenge/card/may-leetcoding-challenge/538/week-5-may-29th-may-31st/3344/
func main() {
	pp.Println("=========================================")
	pp.Println(canFinish(2, [][]int{{1, 0}}))
	pp.Println(true)
	pp.Println("=========================================")
	pp.Println(canFinish(2, [][]int{{1, 0}, {0, 1}}))
	pp.Println(false)
	pp.Println("=========================================")
	pp.Println(canFinish(1, [][]int{}))
	pp.Println(true)
	pp.Println("=========================================")
}

func canFinish(numCourses int, prerequisites [][]int) bool {
	graph := make([][]int, numCourses)
	inDegrees := make([]int, numCourses)
	for _, prerequisite := range prerequisites {
		graph[prerequisite[0]] = append(graph[prerequisite[0]], prerequisite[1])
		inDegrees[prerequisite[1]]++
	}

	queue := make([]int, 0, numCourses)
	for i, inDegree := range inDegrees {
		if inDegree == 0 {
			queue = append(queue, i)
			numCourses--
		}
	}

	for len(queue) != 0 {
		for _, adj := range graph[queue[0]] {
			inDegrees[adj]--
			if inDegrees[adj] == 0 {
				numCourses--
				if numCourses == 0 {
					return true
				}
				queue = append(queue, adj)
			}
		}
		queue = queue[1:]
	}

	return numCourses == 0
}

File: 207/course_schedule.swift
// 207. Course Schedule
// https://leetcode.com/problems/course-schedule/
// import Foundation

func main() {
  let solution = Solution()
  print("=========================================")
  print(solution.canFinish(2, [[1, 0]]))
  print(true)
  print("=========================================")
  print(solution.canFinish(2, [[1, 0],[0, 1]]))
  print(false)
  print("=========================================")
}

class Solution {
  func canFinish(_ n: Int, _ edges: [[Int]]) -> Bool {
    return canFinishDFS(n, edges)
    // return canFinishBFS(n, edges)
  }

  func canFinishDFS(_ n: Int, _ edges: [[Int]]) -> Bool {
    var inDegrees = [Int](repeating: 0, count: n)
    var graph = [[Int]](repeating: [Int](), count: n)
    for edge in edges {
      inDegrees[edge[0]] += 1
      graph[edge[1]].append(edge[0])
    }

    var initialVs = [Int]()
    for v in 0..<n {
      if inDegrees[v] == 0 {
        initialVs.append(v)
      }
    }

    var dfs: ((Int) -> Void)!
    dfs = {(v) in
      for u in graph[v] {
        inDegrees[u] -= 1
        if inDegrees[u] == 0 {
          dfs(u)
        }
      }
    }

    for v in initialVs {
      dfs(v)
    }

    for inDegree in inDegrees {
      if inDegree != 0 {
        return false
      }
    }

    return true
  }

  // ［WIP デバッグ中］なにかおかしい
  // 8
  // [[1,0],[2,6],[1,7],[6,4],[7,0],[0,5]]
  func canFinishBFS(_ n: Int, _ edges: [[Int]]) -> Bool {
    var inDegrees = [Int](repeating: 0, count: n)
    var graph = [[Int]](repeating: [Int](), count: n)
    for edge in edges {
      inDegrees[edge[1]] += 1
      graph[edge[0]].append(edge[1])
    }

    var q = Queue<Int>()
    for v in 0..<n {
      if inDegrees[v] == 0 {
        q.enqueue(v)
      }
    }

    if q.isEmpty {
      return false
    }

    while !q.isEmpty {
      let v = q.dequeue()!
      for u in graph[v] {
        inDegrees[u] -= 1
        if inDegrees[u] < 0 {
          return false
        }
        q.enqueue(u)
      }
    }

    return true
  }
}

public struct Queue<T> {
  fileprivate var array = [T]()

  public var isEmpty: Bool {
    return array.isEmpty
  }

  public var count: Int {
    return array.count
  }

  public mutating func enqueue(_ element: T) {
    array.append(element)
  }

  public mutating func dequeue() -> T? {
    if isEmpty {
      return nil
    } else {
      return array.removeFirst()
    }
  }

  public var front: T? {
    return array.first
  }
}

main()
